### PR TITLE
En 3655/peer processor smart contract for handling

### PR DESCRIPTION
**Pull request by sasurobert**
The last piece of processing staking SC. Moving data from SC to peer state.

### CODE REVIEW

File: process/scToProtocol/stakingToPeer.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+
+	sort.Slice(peersData, func(i, j int) bool {
+		return string(peersData[i].Address) < string(peersData[j].Address)
+	})
+
+	return peersData
+}
+
+// VerifyPeerChanges verifies if peer changes from header is the same as the one created while processing
+func (stp *stakingToPeer) VerifyPeerChanges(peerChanges []block.PeerData) error {
```
Comment (raduchis) [inline process/scToProtocol/stakingToPeer.go]: Not used yet

File: process/scToProtocol/stakingToPeer.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+	}
+	if args.ScDataGetter == nil || args.ScDataGetter.IsInterfaceNil() {
+		return process.ErrNilSCDataGetter
+	}
+
+	return nil
+}
+
+// UpdateProtocol applies changes from staking smart contract to peer state and creates the actual peer changes
+func (stp *stakingToPeer) UpdateProtocol(body block.Body, nonce uint64) error {
```
Comment (raduchis) [inline process/scToProtocol/stakingToPeer.go]: Not used yet.

File: process/smartContract/scDataGetter.go [hunk 3]
```diff
@@ -15,21 +17,27 @@ var maxGasValue = big.NewInt(math.MaxInt64)
...
 
 // NewSCDataGetter returns a new instance of scDataGetter
 func NewSCDataGetter(
-	vm vmcommon.VMExecutionHandler,
+	addrConverter state.AddressConverter,
+	vmContainer process.VirtualMachinesContainer,
 ) (*scDataGetter, error) {
 
-	if vm == nil {
+	if vmContainer == nil || vmContainer.IsInterfaceNil() {
```
Comment (iulianpascalau) [inline process/smartContract/scDataGetter.go]: can use check.IfNil()

File: process/smartContract/scDataGetter.go [hunk 4]
```diff
@@ -38,15 +46,24 @@ func (scdg *scDataGetter) Get(scAddress []byte, funcName string, args ...[]byte)
 	if scAddress == nil {
 		return nil, process.ErrNilScAddress
 	}
+	if len(scAddress) < scdg.addrConverter.AddressLen() {
```
Comment (iulianpascalau) [inline process/smartContract/scDataGetter.go]: Might change `<` into `!=`

File: process/smartContract/scDataGetter.go [hunk 5]
```diff
@@ -38,15 +46,24 @@ func (scdg *scDataGetter) Get(scAddress []byte, funcName string, args ...[]byte)
...
+		return nil, process.ErrInvalidRcvAddr
+	}
 	if len(funcName) == 0 {
 		return nil, process.ErrEmptyFunctionName
 	}
 
 	scdg.mutRunSc.Lock()
 	defer scdg.mutRunSc.Unlock()
 
+	vmType := scAddress[core.NumInitCharactersForScAddress-core.VMTypeLen : core.NumInitCharactersForScAddress]
```
Comment (iulianpascalau) [inline process/smartContract/scDataGetter.go]: Don't we have a func for this? Each time I recheck if the slice is done ok.

File: cmd/node/factory/structs.go [hunk 6]
```diff
@@ -7,6 +7,7 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/process/scToProtocol"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done

File: cmd/node/main.go [hunk 7]
```diff
@@ -4,6 +4,10 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/data"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/main.go]: done

File: data/block/metaBlock.go [hunk 8]
```diff
@@ -134,7 +150,7 @@ func (m *MetaBlock) Load(r io.Reader) error {
 // PeerDataGoToCapn is a helper function to copy fields from a Peer Data object to a PeerDataCapn object
 func PeerDataGoToCapn(seg *capn.Segment, src *PeerData) capnp.PeerDataCapn {
 	dest := capnp.AutoNewPeerDataCapn(seg)
-	value, _ := src.Value.GobEncode()
+	value, _ := src.ValueChange.GobEncode()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/metaBlock.go]: done

File: data/block/metaBlock_test.go [hunk 9]
```diff
@@ -12,10 +12,10 @@ import (
 
 func TestPeerData_SaveLoad(t *testing.T) {
 	pd := block.PeerData{
-		PublicKey: []byte("public key"),
-		Action:    block.PeerRegistrantion,
-		TimeStamp: uint64(1234),
-		Value:     big.NewInt(1),
+		PublicKey:   []byte("public key"),
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/metaBlock_test.go]: done

File: dataRetriever/interface.go [hunk 10]
```diff
@@ -1,6 +1,7 @@
 package dataRetriever
 
 import (
+	"github.com/ElrondNetwork/elrond-go/data"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/interface.go]: done

File: integrationTests/vm/iele/vmAgar_test.go [hunk 11]
```diff
@@ -4,6 +4,8 @@ import (
 	"crypto/rand"
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/mock"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/vm/iele/vmAgar_test.go]: done

File: integrationTests/vm/mockVM/vmGet_test.go [hunk 12]
```diff
@@ -3,6 +3,8 @@ package mockVM
 import (
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/mock"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/vm/mockVM/vmGet_test.go]: done

File: process/block/metablock.go [hunk 13]
```diff
@@ -2,6 +2,7 @@ package block
 
 import (
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/node/external"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/smartContract/scDataGetter.go [hunk 14]
```diff
@@ -2,6 +2,8 @@ package smartContract
 
 import (
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/core"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/scDataGetter.go]: done

File: cmd/node/factory/structs.go [hunk 15]
```diff
@@ -1956,6 +1977,24 @@ func newMetaBlockProcessor(
 		return nil, err
 	}
 
+	scDataGetter, err := smartContract.NewSCDataGetter(state.AddressConverter, vmContainer)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: processing is done sync, it should be fine.

File: cmd/node/factory/structs.go [hunk 16]
```diff
@@ -1956,6 +1977,24 @@ func newMetaBlockProcessor(
...
+		Hasher:      core.Hasher,
+		Marshalizer: core.Marshalizer,
+		//TODO: we need actual peer state here
+		PeerState:    state.AccountsAdapter,
+		BaseState:    state.AccountsAdapter,
+		ArgParser:    argsParser,
+		CurrTxs:      data.MetaDatapool.CurrentBlocksTxs(),
+		ScDataGetter: scDataGetter,
+	}
+	smartContractToProt, err := scToProtocol.NewStakingToPeer(argsStaking)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done

File: cmd/node/factory/structs.go [hunk 17]
```diff
@@ -1956,6 +1977,24 @@ func newMetaBlockProcessor(
...
+
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done.

File: cmd/node/factory/structs.go [hunk 18]
```diff
@@ -1973,8 +2012,11 @@ func newMetaBlockProcessor(
 		TxCoordinator:         txCoordinator,
 	}
 	arguments := block.ArgMetaProcessor{
-		ArgBaseProcessor: argumentsBaseProcessor,
-		DataPool:         data.MetaDatapool,
+		ArgBaseProcessor:   argumentsBaseProcessor,
+		DataPool:           data.MetaDatapool,
+		SCDataGetter:       scDataGetter,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: SC is from smart contract.

File: cmd/node/main.go [hunk 19]
```diff
@@ -1392,12 +1390,47 @@ func startStatisticsMonitor(file *os.File, config config.ResourceStatsConfig, lo
...
+		Accounts:         accnts,
+		AddrConv:         addrConv,
+		StorageService:   storageService,
+		BlockChain:       blockChain,
+		ShardCoordinator: shardCoordinator,
+		Marshalizer:      marshalizer,
+		Uint64Converter:  uint64Converter,
+	}
+
+	if shardCoordinator.SelfId() == sharding.MetachainShardId {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/main.go]: it return error if shardID was wrong after. line 1623

File: dataRetriever/dataPool/currentBlockPool.go [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+	c.mutTxs.Unlock()
+}
+
+// GetTx returns the element saved for the hash
+func (c *currentBlockPool) GetTx(txHash []byte) (data.TransactionHandler, error) {
+	c.mutTxs.RLock()
+	defer c.mutTxs.RUnlock()
+
+	tx, ok := c.txsForBlock[string(txHash)]
+	if !ok {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/dataPool/currentBlockPool.go]: added verification on add.

File: dataRetriever/dataPool/currentBlockPool.go [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+// AddTx writes the tx to the map
+func (c *currentBlockPool) AddTx(txHash []byte, tx data.TransactionHandler) {
+	c.mutTxs.Lock()
+	c.txsForBlock[string(txHash)] = tx
+	c.mutTxs.Unlock()
+}
+
+// IsInterfaceNil returns true if underlying object is nil
+func (c *currentBlockPool) IsInterfaceNil() bool {
+	if c == nil {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/dataPool/currentBlockPool.go]: done

File: dataRetriever/interface.go [hunk 22]
```diff
@@ -197,6 +198,14 @@ type Uint64SyncMapCacher interface {
 	IsInterfaceNil() bool
 }
 
+// TxsForCurrentBlockHandler defines the methods for the local cacher, info for current round
+type TxsForCurrentBlockHandler interface {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/interface.go]: done

File: dataRetriever/interface.go [hunk 23]
```diff
@@ -207,6 +216,7 @@ type PoolsHolder interface {
 	MiniBlocks() storage.Cacher
 	PeerChangesBlocks() storage.Cacher
 	MetaBlocks() storage.Cacher
+	CurrentBlocksTxs() TxsForCurrentBlockHandler
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/interface.go]: done

File: process/mock/scDataGetterMock.go [hunk 24]
```diff
@@ -0,0 +1,20 @@
+package mock
+
+type ScDataGetterMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/mock/scDataGetterMock.go]: done

File: process/mock/txForCurrentBlockMock.go [hunk 25]
```diff
@@ -0,0 +1,35 @@
+package mock
+
+import "github.com/ElrondNetwork/elrond-go/data"
+
+type TxForCurrentBlockMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/mock/txForCurrentBlockMock.go]: done

File: process/mock/scToProtocolHandlerMock.go [hunk 26]
```diff
@@ -0,0 +1,21 @@
+package mock
+
+import "github.com/ElrondNetwork/elrond-go/data/block"
+
+type SCToProtocolMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/mock/scToProtocolHandlerMock.go]: done

File: process/scToProtocol/stakingToPeer.go [hunk 27]
```diff
@@ -0,0 +1,355 @@
...
+	stp.peerChanges = make(map[string]block.PeerData)
+	stp.mutPeerChanges.Unlock()
+
+	affectedStates, err := stp.getAllModifiedStates(body)
+	if err != nil {
+		return err
+	}
+
+	for key := range affectedStates {
+		adrSrc, err := stp.adrConv.CreateAddressFromPublicKeyBytes([]byte(key))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go]: done

File: process/scToProtocol/stakingToPeer.go [hunk 28]
```diff
@@ -0,0 +1,355 @@
...
+		// no data under key -> peer can be deleted from trie
+		if len(data) == 0 {
+			err = stp.peerUnregistered(peerAcc, nonce)
+			if err != nil {
+				return err
+			}
+			return stp.peerState.RemoveAccount(adrSrc)
+		}
+
+		var stakingData systemSmartContracts.StakingData
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go]: done.

File: process/scToProtocol/stakingToPeer.go [hunk 29]
```diff
@@ -0,0 +1,355 @@
...
+) error {
+	stp.mutPeerChanges.Lock()
+	defer stp.mutPeerChanges.Unlock()
+
+	actualPeerChange := block.PeerData{
+		Address:     account.Address,
+		PublicKey:   account.BLSPublicKey,
+		Action:      0,
+		TimeStamp:   nonce,
+		ValueChange: big.NewInt(0),
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go]: it is value delta.

File: process/smartContract/scDataGetter.go (hunk 5 above)
Comment (sasurobert) [reply process/smartContract/scDataGetter.go]: Done. moved to core.

File: dataRetriever/mock/txForCurrentBlockMock.go [hunk 30]
```diff
@@ -0,0 +1,35 @@
+package mock
+
+import "github.com/ElrondNetwork/elrond-go/data"
+
+type TxForCurrentBlockMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/mock/txForCurrentBlockMock.go]: renamed

File: integrationTests/mock/scDataGetterMock.go [hunk 31]
```diff
@@ -0,0 +1,20 @@
+package mock
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/mock/scDataGetterMock.go]: renamed.

File: integrationTests/vm/iele/vmAgar_test.go [hunk 32]
```diff
@@ -4,14 +4,17 @@ import (
 	"crypto/rand"
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/integrationTests"
```
Comment (raduchis) [inline integrationTests/vm/iele/vmAgar_test.go]: imports are not sorted

File: integrationTests/vm/mockVM/vmGet_test.go [hunk 33]
```diff
@@ -3,21 +3,28 @@ package mockVM
 import (
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/integrationTests"
```
Comment (raduchis) [inline integrationTests/vm/mockVM/vmGet_test.go]: imports are not sorted

File: cmd/node/factory/structs.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/node/factory/structs.go]: We should use camel case

File: cmd/node/main.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/node/main.go]: Not done

File: dataRetriever/mock/txForCurrentBlockMock.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply dataRetriever/mock/txForCurrentBlockMock.go]: The file should also be renamed

File: process/mock/scToProtocolHandlerMock.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/mock/scToProtocolHandlerMock.go]: The file should also be renamed

File: integrationTests/mock/scDataGetterMock.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply integrationTests/mock/scDataGetterMock.go]: Not renamed. Should rename also the file

File: process/mock/scDataGetterMock.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/mock/scDataGetterMock.go]: Not renamed. Should rename also the file

File: process/mock/txForCurrentBlockMock.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/mock/txForCurrentBlockMock.go]: Should also rename the file

File: process/scToProtocol/stakingToPeer.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/scToProtocol/stakingToPeer.go]: Breaking clean code rules. If one function does a lot of stuff will be very hard to understand and maintain

File: process/scToProtocol/stakingToPeer.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/scToProtocol/stakingToPeer.go]: ok

**Pull request merged**